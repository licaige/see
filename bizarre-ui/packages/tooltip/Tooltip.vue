<script setup lang="ts">
import { ref } from 'vue'
import type { CSSProperties } from 'vue'
import { rafTimeout, cancelRaf } from '../index'
interface Props {
  maxWidth?: number // 提示框内容最大宽度，单位px
  content?: string // 展示的文本 string | slot
  title?: string // 提示的文本 string | slot
  fontSize?: number // 提示文本字体大小，单位px，优先级高于 overlayStyle
  color?: string // 提示文本字体颜色，优先级高于 overlayStyle
  backgroundColor?: string // 提示框背景颜色，优先级高于 overlayStyle
  overlayStyle?: CSSProperties // 提示框内容区域样式
}
withDefaults(defineProps<Props>(), {
  maxWidth: 120,
  content: '暂无内容',
  title: '暂无提示',
  fontSize: 14,
  color: '#FFF',
  backgroundColor: 'rgba(0,0,0,.85)',
  overlayStyle: () => ({})
})
const visible = ref(false)
const hideTimer = ref()
const top = ref(0) // 提示框top定位
const left = ref(0) // 提示框left定位
const contentRef = ref() // 声明一个同名的模板引用
const titleRef = ref() // 声明一个同名的模板引用
function getPosition () {
  const contentWidth = contentRef.value.offsetWidth // 内容宽度
  const titleWidth = titleRef.value.offsetWidth // 提示文本宽度
  const titleHeight = titleRef.value.offsetHeight // 提示文本高度
  top.value = titleHeight
  left.value = (titleWidth - contentWidth) / 2
}
const emit = defineEmits(['openChange'])
function onShow () {
  getPosition()
  cancelRaf(hideTimer.value)
  visible.value = true
  emit('openChange', visible.value)
}
function onHide (): void {
  hideTimer.value = rafTimeout(() => {
    visible.value = false
    emit('openChange', visible.value)
  }, 100)
}
</script>
<template>
  <div class="m-tooltip" @mouseenter="onShow" @mouseleave="onHide">
    <div
      ref="titleRef"
      class="m-title"
      :class="{'show-tip': visible}"
      @mouseenter="onShow"
      @mouseleave="onHide"
      :style="`max-width: ${maxWidth}px; top: ${-top}px; left: ${-left}px;`">
      <div
        class="u-title"
        :style="[overlayStyle, `font-size: ${fontSize}px; color: ${color}; background-color: ${backgroundColor};`]">
        <slot name="title">{{ title }}</slot>
      </div>
      <div class="m-arrow">
        <span class="u-arrow" :style="`background-color: ${backgroundColor};`"></span>
      </div>
    </div>
    <div ref="contentRef" class="u-content">
      <slot>{{ content }}</slot>
    </div>
  </div>
</template>
<style lang="less" scoped>
.m-tooltip {
  position: relative;
  display: inline-block;
  .m-title {
    position: absolute;
    z-index: 999;
    width: max-content;
    padding-bottom: 16px;
    pointer-events: none;
    opacity: 0;
    transform-origin: 50% 75%;
    transform: scale(.8); // 缩放变换
    -ms-transform: scale(.8); /* IE 9 */
    -webkit-transform: scale(.8); /* Safari and Chrome */
    transition: transform .25s, opacity .25s;
    .u-title {
      padding: 6px 8px;
      margin: 0 auto;
      text-align: start;
      word-break: break-all;
      box-shadow: 0 3px 12px 0 rgba(0, 0, 0, .08), 0 3px 6px -4px rgba(0, 0, 0, .12), 0 6px 24px 6px rgba(0, 0, 0, .05);
      border-radius: 6px;
      font-weight: 400;
      line-height: 1.5;
    }
    .m-arrow {
      position: absolute;
      z-index: 9;
      bottom: 16px;
      left: 50%;
      transform: translate(-50%, 100%);
      width: 18px;
      height: 10px;
      overflow: hidden;
      .u-arrow {
        position: absolute;
        left: 50%;
        top: 0px;
        transform: translate(-50%, -50%) rotate(45deg);
        margin: 0 auto;
        width: 12px;
        height: 12px;
        border-radius: 0 0 3px 0;
        z-index: 8;
        box-shadow: 2px 2px 3px rgba(0, 0, 0, .05);
      }
    }
  }
  .show-tip {
    pointer-events: auto;
    opacity: 1;
    transform: scale(1); // 缩放变换
    -ms-transform: scale(1); /* IE 9 */
    -webkit-transform: scale(1); /* Safari and Chrome */
  }
  .u-content {
    cursor: pointer;
    font-size: 14px;
    line-height: 1.5;
    color: rgba(0,0,0,.85);
  }
}
</style>
