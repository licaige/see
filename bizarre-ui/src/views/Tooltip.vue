<script setup lang="ts">
function openChange (visible: boolean) {
  console.log('visible:', visible)
}
</script>
<template>
  <div class="ml120">
    <h1>Tooltip 文字提示</h1>
    <h2 class="mt30 mb10">基本使用</h2>
    <Space :size="30">
      <Tooltip :maxWidth="240" @open-change="openChange">
        <template #title>特斯拉(Tesla)是美国一家电动汽车及能源公司，总部位于帕洛阿托(Palo Alto)，市值达2100亿美元，产销电动汽车、太阳能板、及储能设备</template>
        <Button type="primary">特斯拉</Button>
      </Tooltip>
      <Tooltip :maxWidth="380" @open-change="openChange">
        <template #title>《哥斯拉》由传奇影业、华纳兄弟影业公司等联合出品。该片讲述了世界各地异常自然灾害的发生频率的上升，被隐瞒了几十年的秘密逐渐浮出水面，人类即将面临一场来自远古的浩劫，三只强大怪兽也将轮番登场掀起层层高潮，这场混战也将在怪兽与人类之间打响的故事</template>
        <Button type="primary">哥斯拉</Button>
      </Tooltip>
    </Space>
    <h2 class="mt30 mb10">自定义样式</h2>
    <Tooltip
      :maxWidth="320"
      :fontSize="18"
      color="rgba(0, 0, 0, 0.85)"
      backgroundColor="#FFF"
      :overlayStyle="{padding: '12px 18px', borderRadius: '12px'}">
      <template #title>《哥斯拉大战金刚》是由美国传奇影业公司出品。该片讲述了人类计划将所有巨兽从地球上抹去，而传说中哥斯拉和金刚两个王者被设计进行了对决，最终两大巨兽联手破坏人类计划的故事。</template>
      <Button type="primary">哥斯拉大战金刚</Button>
    </Tooltip>
    <h2 class="mt30 mb10">暂无数据</h2>
    <Tooltip>
      <Button type="primary">暂无数据</Button>
    </Tooltip>
  </div>
</template>
