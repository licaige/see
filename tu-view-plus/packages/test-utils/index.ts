export { default as defineGetter } from './define-getter';
export { default as makeMount } from './make-mount';
export { default as tick } from './tick';
