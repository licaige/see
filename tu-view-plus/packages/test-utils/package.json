{
  "name": "@tu-view-plus/test-utils",
  "version": "1.0.0",
  "private": true,
  "main": "index.js",
  "license": "MIT"
}