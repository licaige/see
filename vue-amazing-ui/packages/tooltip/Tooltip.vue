<script setup lang="ts">
import { ref } from 'vue'
import type { CSSProperties } from 'vue'
import { rafTimeout, cancelRaf } from '../index'
interface Props {
  maxWidth?: number // 提示框内容最大宽度，单位px
  content?: string // 展示的文本 string | slot
  tooltip?: string // 提示的文本 string | slot
  fontSize?: number // 提示文本字体大小，单位px，优先级高于 overlayStyle
  color?: string // 提示文本字体颜色，优先级高于 overlayStyle
  backgroundColor?: string // 提示框背景颜色，优先级高于 overlayStyle
  overlayStyle?: CSSProperties // 提示框内容区域样式
}
withDefaults(defineProps<Props>(), {
  maxWidth: 120,
  content: '暂无内容',
  tooltip: '暂无提示',
  fontSize: 14,
  color: '#FFF',
  backgroundColor: 'rgba(0, 0, 0, .85)',
  overlayStyle: () => ({})
})
const visible = ref(false)
const hideTimer = ref()
const top = ref(0) // 提示框top定位
const left = ref(0) // 提示框left定位
const contentRef = ref() // 声明一个同名的模板引用
const tooltipRef = ref() // 声明一个同名的模板引用
function getPosition () {
  const contentWidth = contentRef.value.offsetWidth // 展示文本宽度
  const tooltipWidth = tooltipRef.value.offsetWidth // 提示文本宽度
  const tooltipHeight = tooltipRef.value.offsetHeight // 提示文本高度
  top.value = tooltipHeight + 4
  left.value = (tooltipWidth - contentWidth) / 2
}
const emit = defineEmits(['openChange'])
function onShow () {
  getPosition()
  cancelRaf(hideTimer.value)
  visible.value = true
  emit('openChange', visible.value)
}
function onHide (): void {
  hideTimer.value = rafTimeout(() => {
    visible.value = false
    emit('openChange', visible.value)
  }, 100)
}
</script>
<template>
  <div class="m-tooltip" @mouseenter="onShow" @mouseleave="onHide">
    <div
      ref="tooltipRef"
      class="m-tooltip-content"
      :class="{'show-tip': visible}"
      :style="`--tooltip-font-size: ${fontSize}px; --tooltip-color: ${color}; --tooltip-background-color: ${backgroundColor}; max-width: ${maxWidth}px; top: ${-top}px; left: ${-left}px;`"
      @mouseenter="onShow"
      @mouseleave="onHide">
      <div
        class="u-tooltip"
        :style="overlayStyle">
        <slot name="tooltip">{{ tooltip }}</slot>
      </div>
      <div class="m-tooltip-arrow">
        <span class="u-tooltip-arrow"></span>
      </div>
    </div>
    <div ref="contentRef">
      <slot>{{ content }}</slot>
    </div>
  </div>
</template>
<style lang="less" scoped>
.m-tooltip {
  position: relative;
  display: inline-block;
  .m-tooltip-content {
    position: absolute;
    z-index: 999;
    width: max-content;
    padding-bottom: 12px;
    pointer-events: none;
    opacity: 0;
    transform-origin: 50% 75%;
    transform: scale(.8); // 缩放变换
    -ms-transform: scale(.8); /* IE 9 */
    -webkit-transform: scale(.8); /* Safari and Chrome */
    transition: transform .25s, opacity .25s;
    .u-tooltip {
      min-width: 32px;
      min-height: 32px;
      padding: 6px 8px;
      font-size: var(--tooltip-font-size);
      color: var(--tooltip-color);
      line-height: 1.5714285714285714;
      text-align: start;
      text-decoration: none;
      word-wrap: break-word;
      background-color: var(--tooltip-background-color);
      border-radius: 6px;
      box-shadow: 0 6px 16px 0 rgba(0, 0, 0, .08), 0 3px 6px -4px rgba(0, 0, 0, .12), 0 9px 28px 8px rgba(0, 0, 0, .05);
    }
    .m-tooltip-arrow {
      position: absolute;
      z-index: 9;
      left: 50%;
      bottom: 12px;
      transform: translateX(-50%) translateY(100%) rotate(180deg);
      display: block;
      border-radius: 0 0 2px;
      pointer-events: none;
      width: 32px;
      height: 32px;
      overflow: hidden;
      &::before {
        position: absolute;
        bottom: 0;
        inset-inline-start: 0;
        width: 32px;
        height: 8px;
        background: var(--tooltip-background-color);
        clip-path: path('M 6.343145750507619 8 A 4 4 0 0 0 9.17157287525381 6.82842712474619 L 14.585786437626904 1.414213562373095 A 2 2 0 0 1 17.414213562373096 1.414213562373095 L 22.82842712474619 6.82842712474619 A 4 4 0 0 0 25.65685424949238 8 Z');
        content: "";
      }
      &::after {
        position: absolute;
        width: 11.31370849898476px;
        height: 11.31370849898476px;
        bottom: 0;
        inset-inline: 0;
        margin: auto;
        border-radius: 0 0 2px 0;
        transform: translateY(50%) rotate(-135deg);
        box-shadow: 3px 3px 7px rgba(0, 0, 0, .1);
        z-index: 0;
        background: transparent;
        content: "";
      }
    }
  }
  .show-tip {
    pointer-events: auto;
    opacity: 1;
    transform: scale(1); // 缩放变换
    -ms-transform: scale(1); /* IE 9 */
    -webkit-transform: scale(1); /* Safari and Chrome */
  }
}
</style>
