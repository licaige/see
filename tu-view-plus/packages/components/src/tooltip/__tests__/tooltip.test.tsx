import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import Tooltip from '../index';

describe('Tooltip', () => {
  test('render', async () => {
    const positions = ['top', 'tl', 'tr', 'bottom', 'bl', 'br', 'left', 'lt', 'lb', 'right', 'rt', 'rb'];
    const wrapper = mount(
      {
        template: positions.map(item => `
        <tu-tooltip
          position="${item}"
          content="content"
          :default-popup-visible="true"
          :render-to-body="false"
        >
          <button>Button</button>
        </tu-tooltip>
      `).join('/n'),
      },
      {
        global: {
          plugins: [Tooltip],
        },
      }
    );

    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should emit popupVisibleChange event', async () => {
    const wrapper = mount(Tooltip, {
      props: {
        mouseEnterDelay: 0,
        mouseLeaveDelay: 0,
      },
      slots: {
        default: '<button>Button</button>',
        content: 'Content',
      },
    });

    await wrapper.find('button').trigger('mouseenter');
    await wrapper.find('button').trigger('mouseleave');

    expect(wrapper.emitted('popupVisibleChange')).toHaveLength(2);
  });
});
