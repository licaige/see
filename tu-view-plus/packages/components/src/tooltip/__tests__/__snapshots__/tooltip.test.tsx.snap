// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > render 1`] = `
"<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--top tu-tooltip\\" style=\\"z-index: 1001; pointer-events: none;\\" trigger-placement=\\"top\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--tl tu-tooltip\\" style=\\"z-index: 1002; pointer-events: none;\\" trigger-placement=\\"tl\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--tr tu-tooltip\\" style=\\"z-index: 1003; pointer-events: none;\\" trigger-placement=\\"tr\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--bottom tu-tooltip\\" style=\\"z-index: 1004; pointer-events: none;\\" trigger-placement=\\"bottom\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--bl tu-tooltip\\" style=\\"z-index: 1005; pointer-events: none;\\" trigger-placement=\\"bl\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--br tu-tooltip\\" style=\\"z-index: 1006; pointer-events: none;\\" trigger-placement=\\"br\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--left tu-tooltip\\" style=\\"z-index: 1007; pointer-events: none;\\" trigger-placement=\\"left\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--lt tu-tooltip\\" style=\\"z-index: 1008; pointer-events: none;\\" trigger-placement=\\"lt\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--lb tu-tooltip\\" style=\\"z-index: 1009; pointer-events: none;\\" trigger-placement=\\"lb\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--right tu-tooltip\\" style=\\"z-index: 1010; pointer-events: none;\\" trigger-placement=\\"right\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--rt tu-tooltip\\" style=\\"z-index: 1011; pointer-events: none;\\" trigger-placement=\\"rt\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->
 /n
<button>Button</button>
<!--teleport start-->

<div class=\\"tu-popup tu-popup--rb tu-tooltip\\" style=\\"z-index: 1012; pointer-events: none;\\" trigger-placement=\\"rb\\" role=\\"tooltip\\">
  <div class=\\"tu-popup__wrapper tu-zoom-in-fade-out-enter-from tu-zoom-in-fade-out-enter-active\\">
    <div class=\\"tu-popup__content tu-tooltip__content tu-tooltip--medium\\">content</div>
    <div class=\\"tu-popup__arrow tu-popup__arrow--medium tu-tooltip__arrow\\"></div>
  </div>
</div>

<!--teleport end-->"
`;
